File: bot.cpp
#include <QRandomGenerator>

#include "bot.h"

bool isInPlayerBoundingBox(QPoint playerCoord, QVector2D point) {
    return point.x() >= playerCoord.x() && point.x() <= playerCoord.x() + 55 &&
           point.y() >= playerCoord.y() && point.y() <= playerCoord.y() + 95;
}

Bot::Bot(Map *map, Bridge *view, QQuickItem *item, QPoint pos) : Character (map, view, item, pos) {
    m_id = 1;
}

void Bot::goLeft() {
    stopRight();
    Character::goLeft();
}

void Bot::goRight() {
    stopLeft();
    Character::goRight();
}

bool Bot::update() {
    QRandomGenerator qrg = qrg.securelySeeded();state = BotState::Stand;
    switch (state) {
        case BotState::Attack: {
            if (canAttack()) {
                if (qrg.generate() % 10 > 7) {
                    attack(m_map->getMarkedPoint("player").x() + 27, m_map->getMarkedPoint("player").y() + 40);
                    qint32 weapon = qrg.generate() % 20;
                    if (weapon > 10)
                        changeWeapon();
                    if (weapon > 15)
                        changeWeapon();
                    state = BotState::WaitReload;
                } else {
                    if (m_map->getMarkedPoint("player").x() <= m_coord.x()) {
                        goLeft();
                    } else {
                        goRight();
                    }
                    if (not m_map->isFilled(m_boundingBox.translated(0, -60))) {
                        jump();
                    }
                }
            } else if (m_mana < m_weapon->getManaCost()) {
                state = BotState::Flee;
            } else if (isHeroVisible()) {
                // chase player
                if (m_map->getMarkedPoint("player").x() <= m_coord.x()) {
                    goLeft();
                } else {
                    goRight();
                }
            } else {
                state = BotState::Walk;
            }
            break;
        }

        case BotState::Flee: {
            if (m_mana >= 3 * m_weapon->getManaCost()) {
                state = BotState::Attack;
            } else {
                if (m_map->getMarkedPoint("player").x() >= m_coord.x()) {
                    goLeft();
                } else {
                    goRight();
                }
                if (m_goingLeft && m_map->isFilled(m_boundingBox.translated(-5, 0))) {
                    state = BotState::Stand;
                }
                if (m_goingRight && m_map->isFilled(m_boundingBox.translated(5, 0))) {
                    state = BotState::Stand;
                }
            }
            break;
        }

        case BotState::Walk: {
            if (not m_goingRight and qrg.generate() % 2)
                goLeft();
            else if (not m_goingLeft)
                goRight();
            if (not m_map->isFilled(m_boundingBox.translated(0, -60))) {
                jump();
            }
            if (canAttack()) {
                state = BotState::Attack;
            } else {
                if (m_map->isFilled(m_boundingBox.translated(-5, 0))) {
                    goRight();
                } else if (m_map->isFilled(m_boundingBox.translated(5, 0))) {
                    goLeft();
                } else if (qrg.generate() % 70 == 0){
                    if (m_goingLeft) {
                        goRight();
                    } else if (m_goingRight) {
                        goLeft();
                    }
                }
            }
            break;
        }

        case BotState::Stand: {
            stopLeft();
            stopRight();
            if (m_mana >= 2 * m_weapon->getManaCost()) {
                if (canAttack())
                    state = BotState::Attack;
                else
                    state = BotState::Walk;
            }
            break;
        }

        case BotState::WaitReload: {
            stopLeft();
            stopRight();
            if (m_reload == 0) {
                if (canAttack())
                    state = BotState::Attack;
                else if (m_mana < m_weapon->getManaCost())
                    state = BotState::Flee;
                else
                    state = BotState::Walk;
            }
            break;
        }
    }
    return Character::update();
}

bool Bot::isHeroVisible() {
    QVector2D playerCoord = QVector2D(m_map->getMarkedPoint("player")) + QVector2D(27, 47);
    QVector2D position = QVector2D(getHandPosition());
    QVector2D direction = (playerCoord - position).normalized();

    while (not m_map->isFilled(position.toPoint())) {
        position += direction;
        if (isInPlayerBoundingBox(m_map->getMarkedPoint("player"), position)) {
            return true;
        }
    }
    return false;
}

bool Bot::canAttack() {
    return isHeroVisible() && (m_mana >= m_weapon->getManaCost()) && (m_reload == 0);
}

File: src/bot.ts
import {Vec2, add, rectContains, rectTranslated, sub, vec2Normalized} from "./types.ts";
import {Character} from "./character.ts";
import {GameMap} from "./map.ts";
import {World} from "./world.ts";
import {mana_cost} from "./weapon.ts";

enum BotState {
  Attack,
  Flee,
  Walk,
  Stand,
  WaitReload,
}

function isInPlayerBB(playerCoord: Vec2, p: Vec2): boolean {
  return rectContains({...playerCoord, w: 55, h: 95 }, p);
}

export class Bot {
  character: Character;
  private state = BotState.Attack;

  constructor(position: Vec2) {
    this.character = new Character(position, "bot");
  }

  goLeft(): void {
    this.character.stopRight();
    this.character.goLeft();
  }

  goRight(): void {
    this.character.stopLeft();
    this.character.goRight();
  }

  update(w: World): void {
    const map = w.map;
    switch (this.state) {
    case BotState.Attack: {
      if (this.canAttack(map)) {
        if (Math.random() * 10 > 7) {
          const pp = map.getMarkedPoint("player");
          this.character.attack(w, add(pp, {x: 27, y: 40}));
          const weapon = Math.random() * 20;
          if (weapon > 10)
            this.character.changeWeapon();
          if (weapon > 15)
            this.character.changeWeapon();
          this.state = BotState.WaitReload;
        } else {
          const pp = map.getMarkedPoint("player");
          if (pp.x <= this.character.m_coord.x)
            this.goLeft();
          else
            this.goRight();
          if (!map.isFilledRect(rectTranslated(this.character.boundingBox, 0, -60))) {
            this.character.jump(map);
          }
        }
      } else if (this.character.mana < mana_cost[this.character.weaponId]) {
        this.state = BotState.Flee;
      } else if (this.isHeroVisible(map)) {
        const pp = map.getMarkedPoint("player");
        if (pp.x <= this.character.m_coord.x)
          this.goLeft();
        else
          this.goRight();
      } else {
        this.state = BotState.Walk;
      }
      break;
    }

    case BotState.Flee: {
      if (this.character.mana >= Math.min(100, 3 * mana_cost[this.character.weaponId])) {
        this.state = BotState.Attack;
      } else {
        const pp = map.getMarkedPoint("player");
        if (pp.x >= this.character.m_coord.x) this.goLeft();
        else this.goRight();
        if (this.character.m_goingLeft && map.isFilledRect(rectTranslated(this.character.boundingBox, -5, 0))) {
          this.state = BotState.Stand;
        }
        if (this.character.m_goingRight && map.isFilledRect(rectTranslated(this.character.boundingBox, 5, 0))) {
          this.state = BotState.Stand;
        }
      }
      break;
    }

    case BotState.Walk: {
      if (!this.character.m_goingRight && Math.random() * 10 < 5)
        this.goLeft();
      else if (!this.character.m_goingLeft)
        this.goRight();
      if (!map.isFilledRect(rectTranslated(this.character.boundingBox, 0, -60))) {
        this.character.jump(map);
      }
      if (this.canAttack(map)) {
        this.state = BotState.Attack;
      } else {
        if (map.isFilledRect(rectTranslated(this.character.boundingBox, -5, 0))) {
          this.goRight();
        } else if (map.isFilledRect(rectTranslated(this.character.boundingBox, 5, 0))) {
          this.goLeft();
        } else if (Math.floor(Math.random() * 70) === 0) {
          if (this.character.m_goingLeft)
            this.goRight();
          else if (this.character.m_goingRight)
            this.goLeft();
        }
      }
      break;
    }

    case BotState.Stand: {
      this.character.stopLeft();
      this.character.stopRight();
      if (this.character.mana >= 2 * mana_cost[this.character.weaponId]) {
        this.state =
          this.canAttack(map) ? BotState.Attack :
          BotState.Walk;
      }
      break;
    }

    case BotState.WaitReload: {
      this.character.stopLeft();
      this.character.stopRight();
      if (this.character.m_reload === 0) {
        this.state =
          this.canAttack(map) ? BotState.Attack :
          this.character.mana < mana_cost[this.character.weaponId] ? BotState.Flee :
          BotState.Walk
      }
      break;
    }
    }
    this.character.update(map);
  }

  private isHeroVisible(map: GameMap): boolean {
    const pp = map.getMarkedPoint("player");
    const playerCenter = add(pp, {x: 27, y: 47});
    const pos = this.character.getHandPosition();
    const d = sub(playerCenter, pos);
    const len = Math.sqrt(d.x * d.x + d.y * d.y);
    if (len === 0)
      return false;
    const dir = vec2Normalized(d);
    let p = {...pos};
    while (!map.isFilled(p)) {
      p = add(p, dir);
      if (isInPlayerBB(pp, p))
        return true;
    }
    return false;
  }

  private canAttack(map: GameMap): boolean {
    return this.isHeroVisible(map) && this.character.mana >= mana_cost[this.character.weaponId] && this.character.m_reload === 0;
  }
}
